server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,648 @@
+        node.children = children;
+
+        // Construct star-node if required
+        if (!skipStarNodeCreationForDimensions.contains(childDimensionId) && children.size() > 1) {
+            children.put((long) ALL, constructStarNode(startDocId, endDocId, childDimensionId));
+        }
+
+        // Further split on child nodes if required
+        for (TreeNode child : children.values()) {
+            if (child.endDocId - child.startDocId > maxLeafDocuments) {
+                constructStarTree(child, child.startDocId, child.endDocId);
+            }
+        }
+    }
+
+    /**
+     * Constructs non star tree nodes
+     *
+     * @param startDocId  start document id (inclusive)
+     * @param endDocId    end document id (exclusive)
+     * @param dimensionId id of the dimension in the star tree
+     * @return root node with non-star nodes constructed
+     * @throws IOException throws an exception if we are unable to construct non-star nodes
+     */
+    private Map<Long, TreeNode> constructNonStarNodes(int startDocId, int endDocId, int dimensionId) throws IOException {
+        Map<Long, TreeNode> nodes = new HashMap<>();
+        int nodeStartDocId = startDocId;
+        Long nodeDimensionValue = getDimensionValue(startDocId, dimensionId);
+        for (int i = startDocId + 1; i < endDocId; i++) {
+            Long dimensionValue = getDimensionValue(i, dimensionId);
+            if (!dimensionValue.equals(nodeDimensionValue)) {
+                TreeNode child = getNewNode();
+                child.dimensionId = dimensionId;
+                child.dimensionValue = nodeDimensionValue;
+                child.startDocId = nodeStartDocId;
+                child.endDocId = i;
+                nodes.put(nodeDimensionValue, child);
+
+                nodeStartDocId = i;
+                nodeDimensionValue = dimensionValue;
+            }
+        }
+        TreeNode lastNode = getNewNode();
+        lastNode.dimensionId = dimensionId;
+        lastNode.dimensionValue = nodeDimensionValue;
+        lastNode.startDocId = nodeStartDocId;
+        lastNode.endDocId = endDocId;
+        nodes.put(nodeDimensionValue, lastNode);
+        return nodes;
+    }
+
+    /**
+     * Constructs star tree nodes
+     *
+     * @param startDocId  start document id (inclusive)
+     * @param endDocId    end document id (exclusive)
+     * @param dimensionId id of the dimension in the star tree
+     * @return root node with star nodes constructed
+     * @throws IOException throws an exception if we are unable to construct non-star nodes
+     */
+    private TreeNode constructStarNode(int startDocId, int endDocId, int dimensionId) throws IOException {
+        TreeNode starNode = getNewNode();
+        starNode.dimensionId = dimensionId;
+        starNode.dimensionValue = ALL;
+        starNode.isStarNode = true;
+        starNode.startDocId = numStarTreeDocs;
```

Comment:
Shouldn't this be `startDocId` that is passed as an argument?

Reply:
The newly added star documents which are getting created as part of this method will be post the num star tree docs. So that will become the start id and end id will be post we create the star docs.

The recursion still will be on start id to end doc id .